src/AsyncBridge/AsyncCompatLibExtensions.cs

```diff
@@ -85,5 +97,231 @@ public static ConfiguredTaskAwaitable ConfigureAwait(this Task task, bool contin
+    /// canceled already.
+    /// </para>
+    /// </remarks>
+    public static void CancelAfter(this CancellationTokenSource cancelSource, int millisecondsDelay)
+    {
+        if (millisecondsDelay < Timeout.Infinite)
+            throw new ArgumentOutOfRangeException(nameof(millisecondsDelay));
+
+        cancelSource.CancelAfter(new TimeSpan(millisecondsDelay * TimeSpan.TicksPerMillisecond));
+    }
+
+    /// <summary>
+    /// Schedules a Cancel operation on this <see cref="T:System.Threading.CancellationTokenSource"/>.
+    /// </summary>
+    /// <param name="cancelSource">The <see cref="T:System.Threading.CancellationTokenSource"/> to cancel</param>
+    /// <param name="delay">The time span to wait before canceling this <see
+    /// cref="T:System.Threading.CancellationTokenSource"/>.
+    /// </param>
+    /// <exception cref="T:System.ObjectDisposedException">The exception thrown when this <see
+    /// cref="T:System.Threading.CancellationTokenSource"/> has been disposed.
+    /// </exception>
+    /// <exception cref="T:System.ArgumentOutOfRangeException">
+    /// The exception thrown when <paramref name="delay"/> is less than -1 or 
+    /// greater than Int32.MaxValue.
+    /// </exception>
+    /// <remarks>
+    /// <para>
+    /// The countdown for the delay starts during this call.  When the delay expires, 
+    /// this <see cref="T:System.Threading.CancellationTokenSource"/> is canceled, if it has
+    /// not been canceled already.
+    /// </para>
+    /// <para>
+    /// Subsequent calls to CancelAfter will reset the delay for this  
+    /// <see cref="T:System.Threading.CancellationTokenSource"/>, if it has not been
+    /// canceled already.
+    /// </para>
+    /// </remarks>
+    public static void CancelAfter(this CancellationTokenSource cancelSource, TimeSpan delay)
+    {
+        if (cancelSource == null)
+            throw new ArgumentNullException(nameof(cancelSource));
+
+        if (delay < _timeoutInfinite)
+            throw new ArgumentOutOfRangeException(nameof(delay));
+
+        if (cancelSource.IsCancellationRequested)
+            return;
+
+        Timer MyTimer = null;
+
+        // CTS claims it's thread-safe for all methods, so we should be the same.
+        // If someone calls CancelAfter concurrently for the same CTS, we won't explode or cause a memory leak
+        do
+        {
+            if (_cancelTimers.TryGetTimer(cancelSource, out var CurrentTimer))
+            {
+                if (MyTimer != null)
+                    MyTimer.Dispose();
+
+                MyTimer = CurrentTimer;
+                break; // Reset the existing timer to the new delay
+            }
+
+            // An active timer doesn't hold a strong reference to itself, so adding the CTS as state won't hold it alive
+            MyTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
+
+        } while (!_cancelTimers.TryAddTimer(cancelSource, MyTimer));
```

Comment:
In case it ever becomes possible for any call in this loop to throw an exception, and for those who have to deal with thread aborts, can we move `TryAddTimer` into the loop and move the Dispose closer into a finally block like this? Similar to:

```cs
MyTimer = new Timer(OnCancelAfterTimer, cancelSource, Timeout.Infinite, Timeout.Infinite);
var success = false;
try
{
    success = _cancelTimers.TryAddTimer(cancelSource, MyTimer);
}
finally
{
    if (!success) MyTimer.Dispose();
}
if (success) break;
```

Reply:
Actually, I can probably simplify that loop further, since we never remove a timer as long as the CTS is still alive.

Using a try/finally is overkill, IMO. The worst case on an exception is an undisposed Timer that needs finalising, which occurs in the normal running of this code anyway. It'll happen when the delay is set to infinite, or if the CTS is disposed of and GC runs before the Timer elapses.

Sure, we'll clean up the object if ThreadAbort happens during the TryAddTimer call, but there's so many other scenarios where this won't protect us that it doesn't seem worth the extra code complexity. There's no chance for a memory leak here, just a Timer object that might need to wait for finalisation.